slapp/utils/merge_utils.py

```diff
@@ -0,0 +1,237 @@
+from typing import List, Union, Optional, TypedDict, Any
+from pathlib import Path
+import copy
+import jsonlines
+import boto3
+from urllib.parse import urlparse
+import tempfile
+import shutil
+import logging
+import json
+
+from slapp.transfers.utils import read_jsonlines
+
+
+class MergingException(Exception):
+    pass
+
+
+class WorkerAnnotation(TypedDict):
+    workerId: str
+    roiLabel: str
+
+
+class Project(TypedDict):
+    sourceData: str
+    majorityLabel: str
+    workerAnnotations: List[WorkerAnnotation]
+
+
+def compute_majority(labels: list) -> Union[Any, None]:
+    """most prevalent of three labels or None if not 3
+
+    Parameters
+    ----------
+    labels: list
+        items could be for example "cell"/"not cell" or 1/0
+        our SagemakerGroundTruth jobs are using the strings
+
+    Returns
+    -------
+    majority:
+       element of labels that is most prevalent or None if
+       lenght of labels is not 3
+
+    """
+    ulabels = set(labels)
+    if len(ulabels) not in [1, 2]:
+        raise ValueError("`compute_majority()` expects binary classification "
+                         f"but received {len(ulabels)} labels: {ulabels}")
+    counts = {ulabel: labels.count(ulabel) for ulabel in ulabels}
+
+    majority = None
+
+    # require 3 and only 3 annotations
+    if sum(counts.values()) == 3:
+        mcount = max(counts.values())
+        r_counts = {v: k for k, v in counts.items()}
+        majority = r_counts[mcount]
+
+    return majority
+
+
+def get_project_key(record: dict) -> Union[str, None]:
```

Comment:
I'd make a note in the docstring that this looks for an additional key in the data that ends with "-metadata", then strips the "-metadata" to get the project_key. If SageMaker changes the way they format the output then this function will break, so it will be easier to see if we record it in the docstring.

Reply:
done